target/linux/ipq40xx/patches-6.6/999-PCI-qcom-add-hack-compatible-for-ipq4019-Lantiq-DSL.patch

```diff
@@ -0,0 +1,153 @@
++	if (ret)
++		return ret;
++
++	res->lantiq_hack_virt = dma_alloc_coherent(dev, sizeof(u32),
++						   &res->lantiq_hack_phys,
++						   GFP_ATOMIC);
++	if (!res->lantiq_hack_virt) {
++		dev_err(dev, "failed to allocate DMA for lantiq hack\n");
++		return 0;
++	}
++
++	/* Fake Lantiq CPU ID register */
++	val = (u32 *)res->lantiq_hack_virt;
++	*val = QCOM_PCIE_2_4_0_CPU_ID_REG_VAL;
++
++	/* Use highest window index */
++	index = pci->num_ib_windows - 1;
++	if (!pci->num_ib_windows) {
++		dev_err(dev, "No inbound ATU window available for magic\n");
++		return 0;
++	}
++
++	addr = QCOM_PCIE_2_4_0_CPU_ID_REG;
++	phys = res->lantiq_hack_phys;
++
++	/* Make it visible to PCIe devices using address translation unit */
++	ret = dw_pcie_prog_inbound_atu(pci, index, PCIE_ATU_TYPE_MEM,
++				       addr, phys, sizeof(u32));
++	if (ret) {
++		dev_err(dev, "timeout waiting for IATU for lantiq hack: %d\n", ret);
++		return 0;
++	}
++
++	/* Set vendor/device ID of GRX500 PCIe host */
++	dw_pcie_dbi_ro_wr_en(pci);
++	dw_pcie_writew_dbi(pci, PCI_VENDOR_ID, QCOM_PCIE_2_4_0_GRX500_VENDOR_ID);
++	dw_pcie_writew_dbi(pci, PCI_DEVICE_ID, QCOM_PCIE_2_4_0_GRX500_DEVICE_ID);
++	dw_pcie_dbi_ro_wr_dis(pci);
++
++	return 0;
++}
++
+ static int qcom_pcie_get_resources_2_4_0(struct qcom_pcie *pcie)
+ {
+ 	struct qcom_pcie_resources_2_4_0 *res = &pcie->res.v2_4_0;
+@@ -703,6 +769,18 @@ static void qcom_pcie_deinit_2_4_0(struct qcom_pcie *pcie)
+ 	clk_bulk_disable_unprepare(res->num_clks, res->clks);
+ }
+ 
++static void qcom_pcie_deinit_2_4_0_lantiq_hack(struct qcom_pcie *pcie)
++{
++	struct qcom_pcie_resources_2_4_0 *res = &pcie->res.v2_4_0;
++	struct dw_pcie *pci = pcie->pci;
++	struct device *dev = pci->dev;
++
++	dma_free_coherent(dev, sizeof(u32),
++			  res->lantiq_hack_virt, res->lantiq_hack_phys);
++
++	qcom_pcie_deinit_2_4_0(pcie);
++}
++
+ static int qcom_pcie_init_2_4_0(struct qcom_pcie *pcie)
+ {
+ 	struct qcom_pcie_resources_2_4_0 *res = &pcie->res.v2_4_0;
+@@ -1320,6 +1401,15 @@ static const struct qcom_pcie_ops ops_2_4_0 = {
+ 	.ltssm_enable = qcom_pcie_2_3_2_ltssm_enable,
+ };
+ 
++/* Qcom IP rev.: 2.4.0	Synopsys IP rev.: 4.20a Lantiq DSL Hack */
++static const struct qcom_pcie_ops ops_2_4_0_lantiq_hack = {
++	.get_resources = qcom_pcie_get_resources_2_4_0,
++	.init = qcom_pcie_init_2_4_0,
++	.post_init = qcom_pcie_post_init_2_3_2_lantiq_hack,
++	.deinit = qcom_pcie_deinit_2_4_0_lantiq_hack,
++	.ltssm_enable = qcom_pcie_2_3_2_ltssm_enable,
++};
++
+ /* Qcom IP rev.: 2.3.3	Synopsys IP rev.: 4.30a */
+ static const struct qcom_pcie_ops ops_2_3_3 = {
+ 	.get_resources = qcom_pcie_get_resources_2_3_3,
+@@ -1659,6 +1749,7 @@ static const struct of_device_id qcom_pcie_match[] = {
+ 	{ .compatible = "qcom,pcie-apq8064", .data = &cfg_2_1_0 },
+ 	{ .compatible = "qcom,pcie-apq8084", .data = &cfg_1_0_0 },
+ 	{ .compatible = "qcom,pcie-ipq4019", .data = &cfg_2_4_0 },
++	{ .compatible = "qcom,pcie-ipq4019-lantiq-hack", .data = &ops_2_4_0_lantiq_hack },
```

Comment:
I think instead of a separate compatible it would be better to have a property for that, and then just check for the property in `qcom_pcie_post_init_2_3_2()`. `qcom_pcie_deinit_2_4_0()` doesn't even need to check for it, just check for `lantiq_hack_virt` being non-`NULL` (technically you wouldn't even need to do that, as AFAICT `dma_free_coherent()` does nothing if the passed `cpuaddr` (third argument) is `NULL`, though I'm not sure if this is guaranteed behavior).

Reply:
I know it would result in less code delta but with this approach we limit case of rebase or refresh where code gets moved around... If you check the driver, it's very similar and on various refresh (even from upstream maintainers) it did happen that some code moved to the wrong function resulting in all kind of regression.